Nic::BMC CRUD and proxy operations API

dLobatog opened the pull request:
This is an extension to the foreman API that allows users to perform operations on their interfaces, just BMC as of now, but it's easily extensible. In addition to that, it also allows users to create/show/update/delete BMC interfaces associated to hosts, which in turn makes massive installations of physical machines in foreman with these interfaces significantly easier.

Calls should be done to /api/hosts/:host_id_or_name/interfaces as explained in app/controllers/api/v2/interfaces_controller.rb, power operations for instance go like

```
PUT api/hosts/:host_id_or_name/interfaces/:interface_id/power  , { "power_action" : "cycle" }
```

dLobatog committed:
Nic::BMC CRUD and proxy operations API

Minor tabs/spaces fix

Permissions setup

Minor typo in apipie docs

Username and password do not need to be shown on GET operations. POST and PUT operations still set the username and password

API power operations routes and tests are changed

Used host model methods as suggested by Dominic, removed LAN operations

Permissions updated according with the new BMC API

Allow API calls to use ipmitool/freeipmi/ipmi standard power operation syntax in addition to Fog's syntax

Removed fixtures specifications and added them to test helper

Flattened attrs username password to avoid keeping symbols in memory

Sorted out permissions and removed api/v2/interfaces/lan routing

fixture host_id needs to be a small number for Postgres tests to pass

Access permissions fixed (edit hosts have access to create/delete interfaces)

Virt services support new interface methods

Typos on error messages

Used nested objects according to @isratrade refactor

domcleal commented:
@eLobato could you have a look at the tests please?  They all seem to have failed with this:

<pre>
undefined method `primary_key' for Nic:Module (NoMethodError)
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-2/gems/activerecord-3.2.13/lib/active_record/fixtures.rb:715:in `find'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-2/gems/activerecord-3.2.13/lib/active_record/fixtures.rb:802:in `block (5 levels) in setup_fixture_accessors'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-2/gems/activerecord-3.2.13/lib/active_record/identity_map.rb:72:in `without'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-2/gems/activerecord-3.2.13/lib/active_record/fixtures.rb:801:in `block (4 levels) in setup_fixture_accessors'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-2/gems/activerecord-3.2.13/lib/active_record/fixtures.rb:797:in `map'
/usr/local/rvm/gems/ruby-1.9.3-p392@test_develop_pull_request-2/gems/activerecord-3.2.13/lib/active_record/fixtures.rb:797:in `block (3 levels) in setup_fixture_accessors'
/var/lib/workspace/workspace/test_develop_pull_request/database/mysql/puppet/3.0/ruby/1.9.3/test/functional/api/v2/interfaces_controller_test.rb:45:in `block (2 levels) in <class:InterfacesControllerTest>'
</pre>

domcleal commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -0,0 +1,135 @@
+        process_response @interface.destroy
+      end
+
+      api :PUT, "/hosts/:host_id/interfaces/:id/power", "Run power operation on interface. "
```
I'm not sure this should be per interface, should it just be on the host itself?  That's how the UI works.

It'd also make it more transparent between compute resources + BMC then.

domcleal commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -0,0 +1,135 @@
+        end
+      end
+
+      api :PUT, "/hosts/:host_id/interfaces/:id/lan", "Run lan operation on interface."
```
This feels a bit too raw to add straight to the API.

domcleal commented:
I'm not sure making a BMC-specific API is where we want to go, having a more implementation-agnostic API would seem more useful to me, e.g.

```
/host/:id/power
/host/:id/boot
/host/:host_id/interfaces
```

And then /interfaces could be generic, without being tied to BMC interfaces.

What do you think @isratrade?

dLobatog commented:
I generally agree with having something fully generic, but a host might have several interfaces, how could something like /host/:id/power choose which interface to use?

domcleal commented:
Currently `Host::Managed#bmc_nic` uses `interfaces.bmc.first`.  I'm not saying that's correct, but I think it should work in the same way as the UI.

logicminds commented:
The above PUT methods will not work and are unsupported through foreman and smart proxy.  Its also confusing as BMC devices only have one interface.  Even more confusing is that BMC devices can even share the same physical port as the eth0 interface.  Yikes, but very cool if cabling is an issue.

Futhermore, the underlying support library (Rubyipmi) does not allow setting of the BMC interface using TCP.  Although this is possible its currently not supported and probably laden with bugs.  Additionally, the only supported method so far is setting the BMC lan settings when TCP is not used to configure the interface using the openipmi driver.  This is because the smart proxy cannot configure an interface which it cannot connect to.  If you want to massively configure BMC devices, look into openipmi and use in conjunction with rubyipmi or ipmitool.  

Until I get enough free time to create unit tests for bmclib the openipmi / ipmitool is about the only way, which is exactly how the bmclib provider accomplishes this.

dLobatog commented:
@logicminds Maybe the LAN operations should be GET instead, as I can feel how PUT leads people into thinking the LAN settings can be changed. I mostly followed this, for some reason I thought LAN gets were done through a PUT.

http://projects.theforeman.org/projects/smart-proxy/wiki/API

Chassis operations work well though, what's the issue with that? I've tested it on a variety of hardware models with success.

@domcleal So I should use interfaces.bmc.first instead for power/boot/lan ops? I'm totally OK to go forward with that if you want, not sure how will rails handle the routes clashing with host/:id/power for compute resources, especially since payloads are different, etc.. any ideas?

dLobatog commented:
ping? @domcleal  @logicminds

logicminds commented:
@eLobato I don't see why interfaces.bmc needs to be an array.  There will only be one bmc device per compute resource.  However, there may be WLAN functionality down the road.  So I think having something like interfaces.first.provides may be useful method.  Also I haven't see the new BMC support in foreman yet.  So I don't really know what your trying to improve.  So take my comments with a grain of salt.

domcleal commented:
@eLobato I don't think the routes do clash.. the /hosts/:id/power route is only used for the UI, while this would be under /api/hosts/:id/power.

interfaces.bmc is an array because the UI allows you to add any number of network interfaces to a host and doesn't prevent you from adding multiple BMC interfaces for example.

Sorry for the delay in responding!

dLobatog commented:
@domcleal  I just rebased this branch against develop and refactored the routes/tests to what you suggested, I agree it looks nicer, let me know your thoughts

domcleal commented:
@eLobato that looks like a good layout, the only one that sticks out to me is the LAN operations for a BMC NIC.  Maybe that would be better per BMC interface or removed?  I'm not sure if it makes sense to expose this through Foreman as it assumes the BMC device has its LAN configured already for us to access it.. am I missing something?

I haven't done a detailed review yet as tests are failing (follow the Details link), but will add a few comments now.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -12,6 +12,42 @@ def puppetrun
+
+      def power
+        begin
+          render :json => { :power => @host.bmc_nic.proxy.power(:action => params[:power_action]) } , :status => 200
```
You can use `@host.power` here, which supports both BMC and Virt using the PowerManager module.  Note the list of supported actions is longer, see the PowerManager module's SUPPORTED_ACTIONS array.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -12,6 +12,42 @@ def puppetrun
+
+      def boot
+        begin
+          render :json => { :boot => @host.bmc_nic.proxy.boot(:function => 'bootdevice', :device => params[:device]) }, :status => 200
```
Likewise there's an `ipmi_boot` method on `@host`, see `HostsController#ipmi_boot`.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -12,6 +12,42 @@ def puppetrun
+        end
+      end
+
+      api :GET, "/hosts/:host_id/lan", "Run lan operation on interface."
```
As per my comment above, I think this makes less sense on the host or even in the API.

dLobatog replied:
I'll move it back to the Interface, it's accessible via Foreman Proxy by the way (I'm using it)

domcleal replied:
Yeah, though we don't expose everything the proxy does 1:1 into Foreman or its own API, we try to add value on top.. I'm not sure this does.

domcleal commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -0,0 +1,98 @@
+      private
+      attr_reader :nested_obj
+
+      def find_nested_object
```
We have this in a few controllers, maybe it needs to move to the API base controller?  @istrade?

isratrade replied:
yes, I agree that find_nested_object can be refactored and moved to Api::BaseController

domcleal replied:
This has been fixed by @isratrade in #828, so you can rebase @eLobato and get this automatically.  You just need to define a method to say which object types can be nested.

domcleal commented on app/models/nic/base.rb:
```diff
@@ -56,5 +56,10 @@ def uniq_with_hosts
     end
+
+    def symbolize_attrs!
+      self.attrs = self.attrs.symbolize_keys
```
This could be a security risk, we need to avoid symbolising arbitrary input from users as the symbols don't get GCed.

isratrade replied:
@eLobato, ::symbolize_attrs! is what caused test to fail is functional/hosts_controller_test "BMC password updates successful even if attrs serialized field is the only dirty field".  I really have no idea why.  I also don't understand though why are you calling :symbolize_attrs!?

I moved the callback :symbolize_attrs! from before_validation to before_save and the tests passed.  I wish I knew why :)

isratrade replied:
@eLobato, I don't see anything in the logs since the PUT has no parameters, meaning Rails doesn't think it's dirty.  My only guess is that maybe "0" is getting symbolized in parms[:host][:interfaces_attributes["0"][:password] but the callback shouldn't do this ???

dLobatog replied:
I need to symbolize attrs so that POST and PUT requests save the attrs, which is a critical field because it contains the BMC username and password. Otherwise I found attrs are not saved.

domcleal replied:
@eLobato ah, I think I see.  Perhaps instead of a attrs hash, the API should simply be accepting username, password etc as first class parameters, since the NIC::BMC model has setters/getters generated for these automatically.

ohadlevy replied:
agree, the hash is internal, and imho, should not be exposed at all
(remember that yaml vulnerability?)

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -239,16 +239,20 @@
+    map.permission :power_hosts,   {:hosts => [:power],
+                                    :"api/v2/interfaces" => [:power] }
+    map.permission :console_hosts, {:hosts => [:console],
+                                    :"api/v2/interfaces" => [:boot] }
```
This should be in the `ipmi_boot` permission below.

domcleal commented:
What do you think @isratrade?

isratrade commented on test/functional/api/v2/hosts_controller_test.rb:
```diff
@@ -1,6 +1,9 @@
 
 class Api::V2::HostsControllerTest < ActionController::TestCase
+set_fixture_class :nics => Nic::BMC
+fixtures :nics
```
we have already have in test_helper

``` ruby
fixtures :all
```

so fixtures :nics is not needed
I would also move set_fixture_class :nics to test_helper under set_fixture_class({ :hosts => Host::Base })

isratrade commented on test/functional/api/v2/hosts_controller_test.rb:
```diff
@@ -10,4 +13,56 @@ class Api::V2::HostsControllerTest < ActionController::TestCase
+
+    def initialize_proxy_ops
+      User.current = users(:apiadmin)
+      smart_proxies(:bmc)
```
Is this line needed?  What does it do?  Should it be assigned to a variable?

isratrade commented on test/functional/api/v2/interfaces_controller_test.rb:
```diff
@@ -0,0 +1,50 @@
+
+class Api::V2::InterfacesControllerTest < ActionController::TestCase
+  set_fixture_class :nics => Nic::BMC
+  fixtures :nics
```
same as above, don't need "fixtures :nics" and set_fixture_class :nic can be in test_helper

dLobatog commented:
@domcleal @ohadlevy Last update fixes two tests that were not passing (it should go green now), makes username, password and provider first-class parameters (instead of using attrs), and remove the need for symbolizing_attrs, review at will

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -12,6 +12,30 @@ def puppetrun
+      def power
+        begin
+          render :json => { :power => @host.power.send(params[:power_action]) } , :status => 200
+        rescue NoMethodError
```
The data should be validated first before calling send, e.g.

```
    [..] unless PowerManager::SUPPORTED_ACTIONS.include? (params[:power_action])
```

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -12,6 +12,30 @@ def puppetrun
+      def boot
+        begin
+          render :json => { :boot => @host.ipmi_boot(params[:device]) }, :status => 200
+        rescue NoMethodError
```
This should just be a rescue, I don't think it'll throw NoMethodError.  You could get rid of begin/end too and make this method-wide.

domcleal commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -0,0 +1,94 @@
+        end
+      end
+
+      api :PUT, "/hosts/:host_id/interfaces/:id", "Update a nested parameter for host"
```
"nested parameter" should be "interface", same below in DELETE

domcleal commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -0,0 +1,94 @@
+        param :domain_id, Fixnum, :desc => 'Foreman domain id of interface'
+        param :username, String, :required => true
+        param :password, String, :required => true
+        param :provider, String, :required => true, :desc => 'Interface provider, i.e: IPMI'
```
username/password/provider aren't required, are they for non-BMC?  Check PUT too, as it's inconsistent.

domcleal commented on app/services/foreman/access_permissions.rb:
```diff
@@ -239,20 +240,23 @@
                                     :"api/v1/hosts" => [:update],
-                                    :"api/v2/hosts" => [:update]
+                                    :"api/v2/hosts" => [:update],
+                                    :"api/v2/interfaces" => [:update]
```
edit_hosts should probably also get create/delete interfaces permissions

domcleal commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -0,0 +1,94 @@
+      include Api::Version2
+      include Api::TaxonomyScope
+
+      before_filter :find_resource, :only => [:show, :update, :destroy, :power, :boot, :lan]
```
no :lan action anymore

domcleal commented on app/services/power_manager.rb:
```diff
@@ -1,3 +1,4 @@
 module PowerManager
-  SUPPORTED_ACTIONS = [N_('start'), N_('stop'), N_('poweroff'), N_('reboot'), N_('reset'), N_('state')]
+  SUPPORTED_ACTIONS = [N_('start'), N_('stop'), N_('poweroff'), N_('reboot'), N_('reset'), N_('state'),
+                       N_('on'), N_('off'), N_('soft'), N_('cycle'), N_('status')]
```
Doing this breaks the abstraction here as the virt power manager doesn't support these new types.  Is it necessary?

dLobatog replied:
I added support for this on the virt type. Not absolutely necessary, but I've had many people telling me that 'start' 'stop' etc feel very weird for physical boxes, especially since ipmitool  (and all IPMI interfaces) use 'on' 'off' etc..

domcleal replied:
Ok, makes sense I guess to at least permit both as valid inputs.. any objection @ohadlevy?

ohadlevy replied:
none

domcleal commented on app/services/power_manager/virt.rb:
```diff
@@ -22,13 +22,28 @@ def state
     end
 
-    (SUPPORTED_ACTIONS - ['state']).each do |method|
+    (SUPPORTED_ACTIONS - ['state, status']).each do |method|
```
Missing two single quotes inside here, should be two strings

domcleal commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -0,0 +1,94 @@
+      end
+
+      def create
+        interface = @nested_obj.interfaces.new(params[:interface], :without_protection => true)
```
why without_protection on these calls?

dLobatog replied:
Type is a protected attribute and I can't mass assign it or it will throw -

Error Exception: Can't mass-assign protected attributes: type

domcleal replied:
Ack, of course.  I hadn't realised what this was doing before, but now I think we should hide the implementation a little better - perhaps a type parameter that takes known constants (bmc/nic) and instantiates the correct class?  Then we can leave the mass-assignment protection in place, and prevent users creating records of any type they wish.

dLobatog replied:
Where would you put this new 'type' parameter? What constants should it handle? We can possibly just validate known inputs if that's all we need.

domcleal replied:
Good points... let's start with a validation as it's similar to https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/hosts_controller.rb#L32 where we check the resulting class is one of Nic::Base.

domcleal commented on app/controllers/api/v2/hosts_controller.rb:
```diff
@@ -12,6 +12,32 @@ def puppetrun
+        if valid_actions.include? params[:power_action]
+          render :json => { :power => @host.power.send(params[:power_action]) } , :status => 200
+        else
+          render :json => { :error => "NoMethodError: Available methods are #{valid_actions.join(', ')}" }, :status => 422
```
render_error would be better here, also remove `NoMethodError:`.  Ditto for the one in 'boot'.

domcleal commented on app/controllers/api/v2/interfaces_controller.rb:
```diff
@@ -0,0 +1,84 @@
+      end
+
+      def resource_class
+        Nic::BMC
```
Nic::Base?

domcleal commented:
Merged as 8588f9a for Foreman 1.3.  Thanks for the contribution @eLobato!

dLobatog commented:
OK thanks! Did you open a new issue for the type as a string thing?

domcleal commented:
@eLobato no I haven't, but I think it'd be useful if you have a good idea.  We have the same issue in the /api/hosts/facts API where it takes a class name, though in that case, Host::Discovery is a valid option if you have the discovery plugin installed, so it's tricky.

lzap commented:
Note to myself - add as feature (no issue number).

domcleal commented:
@lzap 3046
